apps/web/vibes/soul/primitives/dynamic-form/index.tsx

```diff
@@ -0,0 +1,275 @@
+'use client';
+
+import {
+  FieldMetadata,
+  FormProvider,
+  SubmissionResult,
+  getFormProps,
+  useForm,
+  useInputControl,
+} from '@conform-to/react';
+import { getZodConstraint, parseWithZod } from '@conform-to/zod';
+import { parseAsString, useQueryState, useQueryStates } from 'nuqs';
+import { useActionState, useCallback } from 'react';
+import { useFormStatus } from 'react-dom';
+import { z } from 'zod';
+
+import { ButtonRadioGroup } from '@/vibes/soul/form/button-radio-group';
+import { CardRadioGroup } from '@/vibes/soul/form/card-radio-group';
+import { Checkbox } from '@/vibes/soul/form/checkbox';
+import { Input } from '@/vibes/soul/form/input';
+import { NumberInput } from '@/vibes/soul/form/number-input';
+import { RadioGroup } from '@/vibes/soul/form/radio-group';
+import { Select } from '@/vibes/soul/form/select';
+import { SwatchRadioGroup } from '@/vibes/soul/form/swatch-radio-group';
+import { Button } from '@/vibes/soul/primitives/button';
+
+import { Field, SchemaRawShape, schema } from './schema';
+
+type Action<State, Payload> = (state: Awaited<State>, payload: Payload) => State | Promise<State>;
+
+interface State<F extends Field> {
+  fields: F[];
+  lastResult: SubmissionResult | null;
+}
+
+export type DynamicFormAction<F extends Field> = Action<State<F>, FormData>;
+
+interface Props<F extends Field> {
+  fields: F[];
+  action: DynamicFormAction<F>;
+  syncQueryParams?: boolean;
+  ctaLabel?: string;
+}
+
+export function DynamicForm<F extends Field>({
+  action,
+  fields,
+  ctaLabel = 'Submit',
+  syncQueryParams = false,
+}: Props<F>) {
+  const [params] = useQueryStates(
+    fields.reduce<Record<string, typeof parseAsString>>(
+      (acc, field) => {
+        return { ...acc, [field.name]: parseAsString };
+      },
+      { quantity: parseAsString },
+    ),
+    { shallow: false },
+  );
+  const defaultValue = fields.reduce<{
+    [Key in keyof SchemaRawShape]?: z.infer<SchemaRawShape[Key]>;
+  }>(
+    (acc, field) => ({
+      ...acc,
+      [field.name]: (syncQueryParams ? params[field.name] : null) ?? field.defaultValue ?? '',
+    }),
+    { quantity: 1 },
+  );
+  const [{ lastResult }, formAction] = useActionState(action, {
```

Comment:
If we've got the fields in the action state, should we rename the `field` prop to `initialFields` or `defaultFields` and allow the action to change the fields? For example, imagine an action that adds a new field to the form

Reply:
Fixed.